### DISCUSSION TITLE
How to get extra string/json data by point picking

### DISCUSSION THREAD

**Discussion by magicarm22** [pyvista/pyvista#5988] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?

Hello, pyvista!
Thanks for your amazing tool. I'm creating a tool for QGIS to show 2d map in 3d. For now, I'm working with tubes, and I want to pick a tube to get attributes of this tube (Screenshot №1).
Because I have a lot of tubes, I optimize meshes by combining them to one mesh from one layer.
For now, I need to get the layer, which from it is by clicking on the object. But Picker returns mesh, not my custom class with extra information.

### Description

  This code will contains a lot of QGIS code, but I will try to show only pyvista code
  ``` python
  class Picker:
    def __init__(self):
        self.mesh = None
    
    def __call__(self, picked, *args):
        self.mesh = picked
        # How to get layer from the mesh?
  
  class MultiBlockExtended(pv.MultiBlock):
    def __init__(self, data, layer, **kwargs):
        super().__init__(data, **kwargs)
        self.layer = layer
  
class Window:

    .........................

    def drawPartLine(self, feature, layer, line):
        for i in range(0, len(points) - 1):
            x = points[i].x() - self.min_x
            y = points[i].y() - self.min_y
            z = line.zAt(i) - self.min_z
            x1 = points[i + 1].x() - self.min_x
            y1 = points[i + 1].y() - self.min_y
            z1 = line.zAt(i + 1) - self.min_z
            point1 = np.array([x, y, z])
            point2 = np.array([x1, y1, z1])
            tube = pv.Tube(point1, point2, n_sides=6)
            tubes.append(tube)

        return tubes
        
     def drawLines(self, layer):
        features = list(layer.getFeatures())
        tubes = []
        for feature in features:
            geom = feature.geometry()
            line = geom.constGet()  # line returns a QgsMultiLineString object

            if isinstance(line, QgsMultiLineString):
                for part in line:
                    tubes += self.drawPartLine(feature, layer, part)
            else:
                tubes += self.drawPartLine(feature, layer, line)

        tubes = MultiBlockExtended(tubes).combine()
        self.plotter.add_mesh(tubes, color=color_rgb)
        
   def drawObjects(self):
        for layer in self.selected_layers:
             self.drawLines(layer)
        self.plotter.reset_camera()
        # self.picker = Picker()
        self.plotter.enable_point_picking(self.picker, left_clicking=True, show_message=False, use_mesh=False, picker='cell')
        
drawObjects()  # Just an example. DrawObjects - DrawLines - DrawPartLine.
```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Sat Apr 27 12:10:22 2024 RTZ 2 (зима)

                OS : Windows
            CPU(s) : 8
           Machine : AMD64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : Intel
      GPU Renderer : Intel(R) HD Graphics 4000
       GPU Version : 4.0.0 - Build 10.18.10.4358
  MathText Support : False

  Python 3.9.5 (tags/v3.9.5:0a7dcbd, May  3 2021, 17:27:52) [MSC v.1928 64 bit
  (AMD64)]

           pyvista : 0.43.3
               vtk : 9.3.0
             numpy : 1.26.4
        matplotlib : 3.5.1
            scooby : 0.9.2
             pooch : 1.8.1
            pillow : 9.1.1
         pyvistaqt : 0.11.0
             PyQt5 : 5.15.4
             scipy : 1.6.2
              tqdm : 4.65.0
--------------------------------------------------------------------------------
```

### Screenshots

![Screenshot_3](https://github.com/pyvista/pyvista/assets/28513487/74c50dce-ed3a-450a-bb9e-91a3cf54218c)

**Comment by darikg** [comment]
It looks like you're adding `MultiBlockExtended.combine()` to the plotter, which returns an UnstructuredGrid. Maybe extend that method to store your layer?

```python
class MultiBlockExtended(pv.MultiBlock):
    def __init__(self, data, layer, **kwargs):
        super().__init__(data, **kwargs)
        self.layer = layer

    def combine(self, *args, **kwargs):
        out = super().combine(*args, **kwargs)
        out.layer = self.layer
        return out
```

here I'm just adding an extra python attribute to the UnstructuredGrid. @user27182 was working on some stuff in #5933 to make it easier to store user attributes on meshes.

**Reply by user27182** [reply]
I don't quite follow the example, but if the information you want to store can be serialized as JSON, you could try storing this directly with the mesh with the built-in `user_dict` property instead of storing it as a new `layer` property. E.g. `self.user_dict['layer'] = layer`.

I don't have any knowledge of how the mesh picking works internally, but I suspect if any copying is done internally to return the picked mesh, then the copy likely does not copy python attributes. But, the `user_dict` is stored directly with the mesh as field data, so it should be copied too and returned.

This is a recent unreleased feature though, so you'll need to install the latest dev branch to try it out. Alternatively, if you can reduce the example even more to only include the multiblock, picker, and some example data you want to store rom `layers`, (such that the code be copied and run locally with pyvista only), then I don't mind testing this out on my machine.

**Reply by magicarm22** [reply]
Hello, @darikg and @user27182!
Thanks for your answers! I checked your solution and don't think that I can use them.
1. Add an attribute to the UnrestrictedGrid. This solution looks good, but when I am trying to get information from the Picker class, what can I do, it is only get a mesh. Not an UnrestrictedGrid, but just a `numpy.ndarray`. This is all what I get in the Picker. So I don't have access to the UnrestrictedGrid :(
```python
class Picker:
    def __init__(self):
        self.mesh = None
    def __call__(self, picked, *args):
        self.mesh = picked
        print(type(self.mesh))  <--- `numpy.ndarray`
        print(self.mesh.user_dict['layer_id'])  <---  'numpy.ndarray' object has no attribute 'user_dict'
```

2. Get the data from the user_dict. I installed the newer version, and user_dict works good, I can store layer_id to this dict. But again, I have no access from the Picker class, I just have a `numpy.ndarray`  :(


Thanks for your help! I'll continue to search a solution for my situation

**Comment by darikg** [comment]
Oh right, the default callback just returns the xyz coordinates of the picked point. You need to add `use_picker=True` to the `enable_point_picking` call, and then your picker callback should be

```python
class Picker:
    def __init__(self):
        self.mesh = None

    def __call__(self, picked_point, vtk_picker):
        self.mesh = vtk_picker.GetDataSet()
```
